crates/syntax-bridge/src/lib.rs

```diff
@@ -1043,4 +1053,22 @@ where
     fn error(&mut self, error: String) {
         self.inner.error(error, self.text_pos)
     }
+
+    fn merge_spans(a: SpanData<Ctx>, b: SpanData<Ctx>) -> SpanData<Ctx> {
+        // We don't do what rustc does exactly, rustc does something clever when the spans have different syntax contexts
+        // but this runs afoul of our separation between `span` and `hir-expand`.
+        SpanData {
+            range: if a.ctx == b.ctx {
+                TextRange::new(
+                    std::cmp::min(a.range.start(), b.range.start()),
+                    std::cmp::max(a.range.end(), b.range.end()),
+                )
+            } else {
+                // Combining ranges make no sense when they come from different syntax contexts.
+                a.range
+            },
+            anchor: a.anchor,
+            ctx: a.ctx,
+        }
+    }
```

Comment:
fyi we already have this for the proc-macro server (with a bit more complications to handle the fixup stuff) https://github.com/rust-lang/rust-analyzer/blob/2f55a9155211a9c4a7e061ed4c01bacfa2ebc3d4/crates/proc-macro-srv/src/server_impl/rust_analyzer_span.rs#L348-L375 

Not too relevant (as you pointed out, the separation gives us trouble), just wanted to point it out